File: pkg/models/transaction_item.go
package models

// TransactionItem represents transaction item data stored in database
type TransactionItem struct {
	ItemId          int64  `xorm:"PK"`
	Uid             int64  `xorm:"INDEX(IDX_item_uid_deleted_group_order) NOT NULL"`
	Deleted         bool   `xorm:"INDEX(IDX_item_uid_deleted_group_order) NOT NULL"`
	ItemGroupId     int64  `xorm:"INDEX(IDX_item_uid_deleted_group_order) NOT NULL DEFAULT 0"`
	Name            string `xorm:"VARCHAR(64) NOT NULL"`
	DisplayOrder    int32  `xorm:"INDEX(IDX_item_uid_deleted_group_order) NOT NULL"`
	Hidden          bool   `xorm:"NOT NULL"`
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

// TransactionItemGetRequest represents all parameters of transaction item getting request
type TransactionItemGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

// TransactionItemCreateRequest represents all parameters of transaction item creation request
type TransactionItemCreateRequest struct {
	GroupId int64  `json:"groupId,string"`
	Name    string `json:"name" binding:"required,notBlank,max=64"`
}

// TransactionItemCreateBatchRequest represents all parameters of transaction item batch creation request
type TransactionItemCreateBatchRequest struct {
	Items      []*TransactionItemCreateRequest `json:"items" binding:"required"`
	GroupId    int64                           `json:"groupId,string"`
	SkipExists bool                            `json:"skipExists"`
}

// TransactionItemModifyRequest represents all parameters of transaction item modification request
type TransactionItemModifyRequest struct {
	Id      int64  `json:"id,string" binding:"required,min=1"`
	GroupId int64  `json:"groupId,string"`
	Name    string `json:"name" binding:"required,notBlank,max=64"`
}

// TransactionItemHideRequest represents all parameters of transaction item hiding request
type TransactionItemHideRequest struct {
	Id     int64 `json:"id,string" binding:"required,min=1"`
	Hidden bool  `json:"hidden"`
}

// TransactionItemMoveRequest represents all parameters of transaction item moving request
type TransactionItemMoveRequest struct {
	NewDisplayOrders []*TransactionItemNewDisplayOrderRequest `json:"newDisplayOrders" binding:"required,min=1"`
}

// TransactionItemNewDisplayOrderRequest represents a data pair of id and display order
type TransactionItemNewDisplayOrderRequest struct {
	Id           int64 `json:"id,string" binding:"required,min=1"`
	DisplayOrder int32 `json:"displayOrder"`
}

// TransactionItemDeleteRequest represents all parameters of transaction item deleting request
type TransactionItemDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

// TransactionItemInfoResponse represents a view-object of transaction item
type TransactionItemInfoResponse struct {
	Id           int64  `json:"id,string"`
	Name         string `json:"name"`
	ItemGroupId  int64  `json:"groupId,string"`
	DisplayOrder int32  `json:"displayOrder"`
	Hidden       bool   `json:"hidden"`
}

// FillFromOtherItem fills all the fields in this current item from other transaction item
func (t *TransactionItem) FillFromOtherItem(item *TransactionItem) {
	t.ItemId = item.ItemId
	t.Uid = item.Uid
	t.Deleted = item.Deleted
	t.Name = item.Name
	t.ItemGroupId = item.ItemGroupId
	t.DisplayOrder = item.DisplayOrder
	t.Hidden = item.Hidden
	t.CreatedUnixTime = item.CreatedUnixTime
	t.UpdatedUnixTime = item.UpdatedUnixTime
	t.DeletedUnixTime = item.DeletedUnixTime
}

// ToTransactionItemInfoResponse returns a view-object according to database model
func (t *TransactionItem) ToTransactionItemInfoResponse() *TransactionItemInfoResponse {
	return &TransactionItemInfoResponse{
		Id:           t.ItemId,
		Name:         t.Name,
		ItemGroupId:  t.ItemGroupId,
		DisplayOrder: t.DisplayOrder,
		Hidden:       t.Hidden,
	}
}

// TransactionItemInfoResponseSlice represents the slice data structure of TransactionItemInfoResponse
type TransactionItemInfoResponseSlice []*TransactionItemInfoResponse

// Len returns the count of items
func (s TransactionItemInfoResponseSlice) Len() int {
	return len(s)
}

// Swap swaps two items
func (s TransactionItemInfoResponseSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

// Less reports whether the first item is less than the second one
func (s TransactionItemInfoResponseSlice) Less(i, j int) bool {
	if s[i].ItemGroupId != s[j].ItemGroupId {
		return s[i].ItemGroupId < s[j].ItemGroupId
	}

	return s[i].DisplayOrder < s[j].DisplayOrder
}

File: src/models/transaction_item.ts
export class TransactionItem implements TransactionItemInfoResponse {
    public id: string;
    public name: string;
    public groupId: string;
    public displayOrder: number;
    public hidden: boolean;

    private constructor(id: string, name: string, groupId: string, displayOrder: number, hidden: boolean) {
        this.id = id;
        this.name = name;
        this.groupId = groupId;
        this.displayOrder = displayOrder;
        this.hidden = hidden;
    }

    public toCreateRequest(): TransactionItemCreateRequest {
        return {
            name: this.name,
            groupId: this.groupId
        };
    }

    public toModifyRequest(): TransactionItemModifyRequest {
        return {
            id: this.id,
            groupId: this.groupId,
            name: this.name
        };
    }

    public clone(): TransactionItem {
        return new TransactionItem(this.id, this.name, this.groupId, this.displayOrder, this.hidden);
    }

    public static of(itemResponse: TransactionItemInfoResponse): TransactionItem {
        return new TransactionItem(itemResponse.id, itemResponse.name, itemResponse.groupId, itemResponse.displayOrder, itemResponse.hidden);
    }

    public static ofMulti(itemResponses: TransactionItemInfoResponse[]): TransactionItem[] {
        const items: TransactionItem[] = [];

        for (const itemResponse of itemResponses) {
            items.push(TransactionItem.of(itemResponse));
        }

        return items;
    }

    public static createNewItem(name?: string, groupId?: string): TransactionItem {
        return new TransactionItem('', name || '', groupId || '0', 0, false);
    }
}

export interface TransactionItemCreateRequest {
    readonly groupId: string;
    readonly name: string;
}

export interface TransactionItemCreateBatchRequest {
    readonly items: TransactionItemCreateRequest[];
    readonly groupId: string;
    readonly skipExists: boolean;
}

export interface TransactionItemModifyRequest {
    readonly id: string;
    readonly groupId: string;
    readonly name: string;
}

export interface TransactionItemHideRequest {
    readonly id: string;
    readonly hidden: boolean;
}

export interface TransactionItemMoveRequest {
    readonly newDisplayOrders: TransactionItemNewDisplayOrderRequest[];
}

export interface TransactionItemNewDisplayOrderRequest {
    readonly id: string;
    readonly displayOrder: number;
}

export interface TransactionItemDeleteRequest {
    readonly id: string;
}

export interface TransactionItemInfoResponse {
    readonly id: string;
    readonly name: string;
    readonly groupId: string;
    readonly displayOrder: number;
    readonly hidden: boolean;
}
